File: 0001~0500/0135. Candy/main.py
"""
Source: https://leetcode.com/problems/candy/?envType=study-plan-v2&envId=top-interview-150
Date: 2023/7/7
Skill:
Ref:
Runtime: 183 ms, faster than 30.97%
Memory Usage: 19.1 MB, less than 99.82%
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def candy(self, ratings: List[int]) -> int:
        nums = [1 for _ in range(len(ratings))]
        for i in range(1, len(ratings)):
            if ratings[i] > ratings[i-1]:
                nums[i] = nums[i-1] + 1
        for i in range(len(ratings) - 2, -1, -1):
            if ratings[i] > ratings[i+1]:
                nums[i] = max(nums[i], nums[i+1] + 1)
        return sum(nums)


if __name__ == "__main__":
    s = Solution()

File: 0001~0500/0135. Candy/main.cpp
/**
 * Source: https://rb.gy/sd9lz
 * Date: 2023/9/13
 * Skill:
 * Runtime: 39 ms, faster than 10.34% of C++ online submissions
 * Memory Usage: 20.31 MB, less than 5.14% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>

using namespace std;

#define ll long long

typedef pair<int, int> pairs;

typedef unsigned long long ULL;
typedef pair<ULL, ULL> PULL;


class Solution {
public:
    int candy(vector<int>& ratings) {
        if (ratings.size() == 1)
            return 1;
        vector<int> res(ratings.size(), INT32_MAX);
        priority_queue<pairs, vector<pairs>, greater<pairs>> pq;
        for (int i = 0; i < ratings.size(); i++)
            pq.push(make_pair(ratings[i], i));
        while (!pq.empty()) {
            auto [rating, pos] = pq.top();
            pq.pop();
            if (pos == 0)
                res[pos] = (rating <= ratings[pos + 1]) ? 1 : res[pos + 1] + 1;
            else if (pos == ratings.size() - 1)
                res[pos] = (rating <= ratings[pos - 1]) ? 1 : res[pos - 1] + 1;
            else {
                int low = (ratings[pos + 1] <= ratings[pos - 1]) ? pos + 1 : pos - 1;
                if (rating <= min(ratings[pos + 1], ratings[pos - 1]))
                    res[pos] = 1;
                else if (rating <= max(ratings[pos + 1], ratings[pos - 1]))
                    res[pos] = res[low] + 1;
                else
                    res[pos] = max(res[pos + 1], res[pos - 1]) + 1;
            }
        }
        return accumulate(begin(res), end(res), 0);
    }
};


static const auto io_sync_off = []() {
    // turn off sync
    std::ios::sync_with_stdio(false);
    // untie in/out streams
    std::cin.tie(nullptr);
    return nullptr;
}();

int main() {
    Solution s;
    vector<vector<int>> grid{{0,2,0,0,0,0,0},{0,0,0,2,2,1,0},{0,2,0,0,1,2,0},{0,0,2,2,2,0,2},{0,0,0,0,0,0,0}};
}